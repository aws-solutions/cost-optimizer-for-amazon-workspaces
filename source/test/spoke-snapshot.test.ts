// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as cdk from "aws-cdk-lib";
import { DefaultStackSynthesizer } from "aws-cdk-lib";
import {
  CostOptimizerSpokeStack,
  CostOptimizerSpokeStackProps,
} from "../lib/cost-optimizer-for-amazon-workspaces-spoke-stack";
import { Template } from "aws-cdk-lib/assertions";

export const costOptimizerSpokeStackProperties: CostOptimizerSpokeStackProps = {
  solutionId: "SO0018",
  solutionTradeMarkName: "cost-optimizer-for-amazon-workspaces",
  solutionProvider: "aws",
  solutionBucketName: "solutions",
  solutionName: "workspaces-cost-optimizer",
  solutionVersion: "v2.8.1",
  description:
    "(" +
    "SO0018" +
    ") - The AWS CloudFormation spoke template" +
    " for deployment of the " +
    "workspaces-cost-optimizer" +
    ", Version: " +
    "v2.8.1",
  synthesizer: new DefaultStackSynthesizer({
    generateBootstrapVersionRule: false,
  }),
};

test("spoke stack synth matches the existing snapshot", () => {
  const app = new cdk.App();
  const costOptimizerSpokeStack = new CostOptimizerSpokeStack(
    app,
    "cost-optimizer-for-amazon-workspaces-spoke",
    costOptimizerSpokeStackProperties,
  );
  const template = Template.fromStack(costOptimizerSpokeStack);
  expect(template.toJSON()).toMatchSnapshot();
});
