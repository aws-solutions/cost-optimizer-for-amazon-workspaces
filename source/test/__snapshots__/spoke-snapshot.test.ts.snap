// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spoke stack synth matches the existing snapshot 1`] = `
{
  "Description": "(SO0018) - The AWS CloudFormation spoke template for deployment of the workspaces-cost-optimizer, Version: v2.7.6",
  "Mappings": {
    "Solution": {
      "Data": {
        "AppRegistryApplicationName": "workspaces-cost-optimizer",
        "HubAccountAdminRoleName": "Workspaces-Cost-Optimizer",
        "HubAccountRegistrationFunctionName": "Register-Spoke-Accounts",
        "ID": "SO0018",
        "SpokeAccountWorkspacesRole": "Workspaces-Admin-Spoke",
        "TagKey": "CloudFoundations:CostOptimizerForWorkspaces",
        "Version": "v2.7.6",
      },
    },
  },
  "Metadata": {
    "AWS::CloudFormation::Interface": {
      "ParameterGroups": [
        {
          "Label": {
            "default": "Hub account information",
          },
          "Parameters": [
            "HubAccountId",
          ],
        },
        {
          "Label": {
            "default": "Testing parameters",
          },
          "Parameters": [
            "LogLevel",
          ],
        },
      ],
      "ParameterLabels": {
        "HubAccountId": {
          "default": "Hub account ID",
        },
        "LogLevel": {
          "default": "Logging level",
        },
      },
    },
  },
  "Outputs": {
    "LogLevel": {
      "Export": {
        "Name": "LogLevel",
      },
      "Value": {
        "Ref": "LogLevel",
      },
    },
    "SolutionID": {
      "Export": {
        "Name": "SolutionID",
      },
      "Value": {
        "Fn::FindInMap": [
          "Solution",
          "Data",
          "ID",
        ],
      },
    },
    "SolutionVersion": {
      "Export": {
        "Name": "SolutionVersion",
      },
      "Value": {
        "Fn::FindInMap": [
          "Solution",
          "Data",
          "Version",
        ],
      },
    },
  },
  "Parameters": {
    "HubAccountId": {
      "AllowedPattern": "^\\d{12}$",
      "Description": "The ID of the hub account for the solution. This stack should be deployed in the same region as the hub stack in the hub account.",
      "Type": "String",
    },
    "LogLevel": {
      "AllowedValues": [
        "CRITICAL",
        "ERROR",
        "WARNING",
        "INFO",
        "DEBUG",
      ],
      "Default": "INFO",
      "Type": "String",
    },
  },
  "Resources": {
    "AccountRegistration": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "AccountRegistrationProviderRolePolicy",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "AccountRegistrationProvider",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AccountRegistration",
      "UpdateReplacePolicy": "Delete",
    },
    "AccountRegistrationProvider": {
      "DependsOn": [
        "AccountRegistrationProviderRole",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "The lambda function has access to write logs",
            },
            {
              "id": "W89",
              "reason": "The lambda function does not need access to resources in VPC",
            },
            {
              "id": "W92",
              "reason": "The lambda function only executes on stack creation and deletion and so does not need reserved concurrency.",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Join": [
              "",
              [
                "solutions-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "workspaces-cost-optimizer/v2.7.6/account_registration_provider.zip",
        },
        "Description": "WorkspacesCostOptimizer spoke account registration custom resource provider",
        "Environment": {
          "Variables": {
            "LOG_LEVEL": {
              "Ref": "LogLevel",
            },
            "MANAGEMENT_ROLE_ARN": {
              "Fn::GetAtt": [
                "WorkSpacesManagementRole",
                "Arn",
              ],
            },
            "REGISTER_LAMBDA_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":lambda:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "HubAccountId",
                  },
                  ":function:",
                  {
                    "Fn::FindInMap": [
                      "Solution",
                      "Data",
                      "HubAccountRegistrationFunctionName",
                    ],
                  },
                  "-",
                  {
                    "Ref": "AWS::Region",
                  },
                ],
              ],
            },
            "USER_AGENT_STRING": "AwsSolution/SO0018/v2.7.6",
          },
        },
        "Handler": "account_registration_provider.account_registration_provider.event_handler",
        "Role": {
          "Fn::GetAtt": [
            "AccountRegistrationProviderRole",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Tags": [
          {
            "Key": {
              "Fn::FindInMap": [
                "Solution",
                "Data",
                "TagKey",
              ],
            },
            "Value": {
              "Ref": "AWS::StackName",
            },
          },
        ],
        "Timeout": 300,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "AccountRegistrationProviderRole": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": {
              "Fn::FindInMap": [
                "Solution",
                "Data",
                "TagKey",
              ],
            },
            "Value": {
              "Ref": "AWS::StackName",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AccountRegistrationProviderRolePolicy": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "X-ray requires * policy",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:/",
                    {
                      "Ref": "AWS::Partition",
                    },
                    "/lambda/*",
                  ],
                ],
              },
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":lambda:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "HubAccountId",
                    },
                    ":function:",
                    {
                      "Fn::FindInMap": [
                        "Solution",
                        "Data",
                        "HubAccountRegistrationFunctionName",
                      ],
                    },
                    "-",
                    {
                      "Ref": "AWS::Region",
                    },
                  ],
                ],
              },
            },
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              "-account-registration-provider-role-policy",
            ],
          ],
        },
        "Roles": [
          {
            "Ref": "AccountRegistrationProviderRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WorkSpacesManagementRole": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W28",
              "reason": "Static naming is necessary for hub account to assume this role",
            },
          ],
        },
        "guard": {
          "SuppressedRules": [
            "CFN_NO_EXPLICIT_RESOURCE_NAMES",
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "HubAccountId",
                      },
                      ":role/",
                      {
                        "Fn::FindInMap": [
                          "Solution",
                          "Data",
                          "HubAccountAdminRoleName",
                        ],
                      },
                      "-",
                      {
                        "Ref": "AWS::Region",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::FindInMap": [
                  "Solution",
                  "Data",
                  "SpokeAccountWorkspacesRole",
                ],
              },
              "-",
              {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "Tags": [
          {
            "Key": {
              "Fn::FindInMap": [
                "Solution",
                "Data",
                "TagKey",
              ],
            },
            "Value": {
              "Ref": "AWS::StackName",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WorkSpacesManagementRolePolicy": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "CloudWatch GetMetricStatistics does not support resource level permissions",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "workspaces:DescribeTags",
                "workspaces:DescribeWorkspaces",
                "workspaces:DescribeWorkspaceDirectories",
                "workspaces:ModifyWorkspaceProperties",
                "workspaces:TerminateWorkspaces",
                "workspaces:DescribeWorkspacesConnectionStatus",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":workspaces:*:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":directory/*",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":workspaces:*:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":workspace/*",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":workspaces:*:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":workspacebundle/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudwatch:GetMetricStatistics",
                "cloudwatch:GetMetricData",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              "-workspaces-management-role-policy",
            ],
          ],
        },
        "Roles": [
          {
            "Ref": "WorkSpacesManagementRole",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
